### PR TITLE
Added ws-shell support -- interactive shell version 2

### PR DESCRIPTION
**Pull request by jonasfj**
There is probably still a few bugs in ws-shell, but they should be solved in that module..
This basically adds support for a new protocol for interactive shell.. This protocol uses ack messages, instead of requiring pause/resume messages, which you can't send when network is under pressure.

background: tc-worker will be using a new protocol for interactive shell, because implementing the old (bad) protocol in golang seemed like a bad idea... Eventually, we might migrate docker-worker to use this protocol too.

### CODE REVIEW

File: src/shell/app.jsx [hunk 1]
```diff
@@ -41,8 +41,18 @@ term.onTerminalReady = async () => {
         'exec $SPAWN;'
       ].join('')
     ]
-  });
+  };
+  let client;
+  if (args.v === '1') {
```
Comment (eliperelman) [inline src/shell/app.jsx]: Insert line between declaration and conditionals.

File: src/shell/app.jsx [hunk 2]
```diff
@@ -41,8 +41,18 @@ term.onTerminalReady = async () => {
...
+    client = new DockerExecClient(options);
+    await client.execute();
+    const resize = client.resize;
```
Comment (eliperelman) [inline src/shell/app.jsx]: Insert line before and after this declaration.

File: src/shell/app.jsx [hunk 3]
```diff
@@ -41,8 +41,18 @@ term.onTerminalReady = async () => {
...
+    client.resize = (c, r) => resize.call(client, r, c);
+  } else if (args.v === '2') {
+    client = await wsshell.dial(options);
+  }
 
+  const io = term.io.push();
```
Comment (eliperelman) [inline src/shell/app.jsx]: Insert line after this declaration.

File: src/shell/app.jsx (hunk 1 above)
Comment (jonasfj) [inline src/shell/app.jsx]: before `let client` I can understand

File: src/shell/app.jsx (hunk 2 above)
Comment (eliperelman) [inline src/shell/app.jsx]: Keeps declarations readable from other statements.

Comment (jonasfj) [inline src/shell/app.jsx]: but, but.... const resize is only used inside the `client.resize = (...) => ` statement on the next line... I'm effectively wrapping the resize method to switch argument ordering :)

**Comment by jonasfj** [comment]
Hmm... Can't figure out what's up with npm... will regenerate shrinkwrap and update.

**Pull request merged**